### PR TITLE
offline flag disables `source-repository-package` sync

### PR DESCRIPTION
**Pull request by peterbecich**
Closes: https://github.com/haskell/cabal/issues/9641

How this was tested:
- use the example project: https://github.com/peterbecich/cabal2nix-project-issue
  - this example pulls the `text` library as a `source-repository-package`
- build, install Cabal from this PR
  - `cabal install cabal --overwrite-policy=always`
- in the example project, delete `dist-newstyle` to remove the `text` library cached there
  - `rm -rf dist-newstyle; ~/.cabal/bin/cabal build`
  - `text` library pulled, build succeeds
- next, build with the cached `text` library and `offline` flag
  - `~/.cabal/bin/cabal build  --offline`
  - this succeeds
  - offline flag does not break the cached library
- next, delete `dist-newstyle` again, and attempt build with `--offline`
  - `rm -rf dist-newstyle; ~/.cabal/bin/cabal build --offline`
  - this fails, as intended
  - this proves the offline flag is preventing the `text` library from being pulled
  - the error message could be improved:
```
$ rm -rf dist-newstyle; ~/.cabal/bin/cabal build --offline
.../dist-newstyle/src/text-8e814f5854032da8: getDirectoryContents:openDirStream:
  does not exist (No such file or directory)
```

Please read [Github PR Conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#github-pull-request-conventions) and then fill in *one* of these two templates.

---

**Template Α: This PR modifies `cabal` behaviour**

Include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#other-conventions).
* [x] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [ ] The documentation has been updated, if necessary.
* [x] [Manual QA notes](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#qa-notes) have been included.
* [ ] Tests have been added. (*Ask for help if you don’t know how to write them! Ask for an exemption if tests are too complex for too little coverage!*)

### CODE REVIEW

**Review by geekosaur — APPROVED** [review]

**Comment by geekosaur** [comment]
I do wonder if this should print something (not sure if level info or warning) when skipping due to being offline.

**Comment by peterbecich** [comment]
I have implemented a warning. However, it is still followed by a cryptic error message:
```
$ rm -rf dist-newstyle; ~/.cabal/bin/cabal build --offline
Warning: --offline was specified, skipping sync of repositories:
Warning: https://github.com/haskell/text.git
.../dist-newstyle/src/text-8e814f5854032da8: getDirectoryContents:openDirStream: does not exist (No such file or directory)
```


If the dependency is cached, there is only a warning
```
$ ~/.cabal/bin/cabal build --offline
Warning: --offline was specified, skipping sync of repositories:
Warning: https://github.com/haskell/text.git
Warning: no repository provided none
Resolving dependencies...
```

**Comment by peterbecich** [comment]
@geekosaur , can this be merged? Thanks

**Comment by geekosaur** [comment]
It needs another reviewer; I've attached the needs-review label, which hopefully will get some more eyes on it.

**Comment by peterbecich** [comment]
Thank you

**Review by gbaz — APPROVED** [review]

**Comment by geekosaur** [comment]
I'll add the "merge me" label for you per the [PR guidelines](https://github.com/haskell/cabal/blob/833a17deb0a5c5f99849a5b6bf06b5346105b918/CONTRIBUTING.md#github-pull-request-conventions). There will then be a 2 day cooldown period to let others comment on the PR before Mergify merges it.

**Comment by Kleidukos** [comment]
@mergifyIO backport 3.12

**Pull request merged**
